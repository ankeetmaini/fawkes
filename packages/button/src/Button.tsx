// ref https://tympanus.net/Development/CreativeButtons/

import React, { FunctionComponent, useContext } from 'react';
import styled from '@emotion/styled';
import { ThemeContext } from '@fawkes/core';
import { ThemeType } from '@fawkes/core/types';
import DancingDots from '@fawkes/loader';

type Size = 's' | 'r' | 'l';
interface ButtonProps extends Omit<React.HTMLProps<HTMLButtonElement>, 'size'> {
  loading?: boolean;
  label: string;
  size: Size;
  type?: 'button' | 'submit' | 'reset';
}

const getStyles = (buttonSize: Size, theme: ThemeType) => {
  const factor = {
    s: 0,
    r: 8,
    l: 14,
  };
  const font = {
    s: theme.typography.SMALL.bold,
    r: theme.typography.REGULAR.bold,
    l: theme.typography.LARGE.bold,
  }[buttonSize];

  const padding = {
    s: '6px 10px',
    r: '8px 12px',
    l: '10px 14px',
  }[buttonSize];
  const { size } = font;
  return { padding, size: size + factor[buttonSize] };
};
const Button: FunctionComponent<ButtonProps> = (props: ButtonProps) => {
  const theme = useContext(ThemeContext);
  const { disabled, label, size, loading, ...rest } = props;
  const { padding, size: fontSize } = getStyles(size, theme);
  const StyledButton = styled.button`
    display: inline-block;
    position: relative;
    padding: ${padding};
    margin: 2px 0;
    font-size: ${fontSize}px;
    color: ${theme.colorScheme.secondary.base};
    background: ${theme.colorScheme.primary.base};
    cursor: ${disabled ? 'not-allowed' : 'default'};
    opacity: ${disabled ? 0.7 : 1};
    z-index: 0;
    border-radius: 4px;
    &:after {
      display: block;
      position: absolute;
      z-index: -1;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      content: '';
      background: ${theme.colorScheme.secondary.base};
      transform-origin: 50% 50%;
      transition: transform 0.2s, opacity 0.2s;
      transform: scaleX(0);
      opacity: 0;
    }
    &:hover {
      color: ${theme.colorScheme.primary.base};
      &:after {
        transform: scaleX(0.8);
        opacity: 1;
      }
    }
    &:active {
      &:after {
        transform: scaleX(1);
      }
    }
  `;
  return (
    <StyledButton disabled={disabled} type="button" {...rest}>
      <span
        style={{
          display: 'inline-flex',
          alignItems: 'center',
          justifyContent: 'center',
        }}
      >
        <span>{loading && <DancingDots dotCount={3} />}</span>{' '}
        <span>{label}</span>
      </span>
    </StyledButton>
  );
};

export default Button;
