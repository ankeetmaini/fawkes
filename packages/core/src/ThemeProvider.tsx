import React, { FunctionComponent } from 'react';
import { Color } from '../types/index';
import Theme from './index';

interface ProviderProps {
  customColorScheme?: Color;
}

const { colorScheme, createTypography } = Theme;
const initialContext = {
  colorScheme,
  typography: createTypography(colorScheme),
};

const ThemeContext = React.createContext(initialContext);

const ThemeProvider: FunctionComponent<ProviderProps> = ({
  customColorScheme,
  children,
}) => {
  const themeContext = customColorScheme
    ? {
        colorScheme: customColorScheme,
        typography: createTypography(customColorScheme),
      }
    : initialContext;
  return (
    <ThemeContext.Provider value={themeContext}>
      {children}
    </ThemeContext.Provider>
  );
};

export default ThemeProvider;
