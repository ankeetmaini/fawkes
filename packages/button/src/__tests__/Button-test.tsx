import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup, fireEvent } from '@testing-library/react';
import ThemeProvider from '@fawkes/core';
import { matchers } from 'jest-emotion';
import Button from '../index';

expect.extend(matchers);
afterEach(cleanup);

describe('button', () => {
  test('button renders', async () => {
    const { asFragment } = render(
      <ThemeProvider>
        <Button label="Submit" size="l"></Button>
      </ThemeProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('button hover animation works', () => {
    const tree = renderer
      .create(
        <ThemeProvider>
          <Button label="Submit" size="l"></Button>
        </ThemeProvider>,
      )
      .toJSON();

    // FIXME why is this not working?
    // button's :after element to have transform: scaleX(0)
    // expect(tree).toHaveStyleRule('transform', 'scaleX(0)', {
    //   target: ':after',
    // });

    expect(tree).toHaveStyleRule('transform', 'scaleX(1)', {
      target: ':active:after',
    });

    expect(tree).toHaveStyleRule('transform', 'scaleX(0.8)', {
      target: ':hover:after',
    });
  });

  test('props are propagated', async () => {
    const onClick = jest.fn();
    const onFocus = jest.fn();
    const onBlur = jest.fn();
    const { getByText } = render(
      <ThemeProvider>
        <Button
          label="Submit"
          size="l"
          onClick={onClick}
          onBlur={onBlur}
          onFocus={onFocus}
        ></Button>
      </ThemeProvider>,
    );
    const button = await getByText(/Submit/);
    fireEvent.click(button);
    fireEvent.focus(button);
    fireEvent.blur(button);

    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onFocus).toHaveBeenCalledTimes(1);
  });
});
