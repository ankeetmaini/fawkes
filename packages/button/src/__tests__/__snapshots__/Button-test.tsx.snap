// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button button renders large 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: Roboto;
  color: #FFFFFF;
}

.emotion-1 {
  display: inline-block;
  padding: 6px 18px;
  width: 94px;
  height: 32px;
  font-size: 16px;
  background: #6979F8;
  cursor: auto;
  border-radius: 3px;
  border: none;
  -webkit-transition: -webkit-transform 0.1s;
  -webkit-transition: transform 0.1s;
  transition: transform 0.1s;
}

.emotion-1:active:not([disabled]) {
  -webkit-transform: translateY(3px);
  -ms-transform: translateY(3px);
  transform: translateY(3px);
}

<button
    class="emotion-1"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Submit
    </span>
  </button>
</DocumentFragment>
`;

exports[`button button renders regular 1`] = `
<DocumentFragment>
  .emotion-0 {
  font-family: Roboto;
  color: #FFFFFF;
}

.emotion-1 {
  display: inline-block;
  padding: 6px 18px;
  width: 90px;
  height: 32px;
  font-size: 14px;
  background: #6979F8;
  cursor: auto;
  border-radius: 3px;
  border: none;
  -webkit-transition: -webkit-transform 0.1s;
  -webkit-transition: transform 0.1s;
  transition: transform 0.1s;
}

.emotion-1:active:not([disabled]) {
  -webkit-transform: translateY(3px);
  -ms-transform: translateY(3px);
  transform: translateY(3px);
}

<button
    class="emotion-1"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Submit
    </span>
  </button>
</DocumentFragment>
`;

exports[`button button renders small 1`] = `
<DocumentFragment>
  .emotion-1 {
  display: inline-block;
  padding: 4px 16px;
  width: 86px;
  height: 24px;
  font-size: 12px;
  background: #6979F8;
  cursor: auto;
  border-radius: 3px;
  border: none;
  -webkit-transition: -webkit-transform 0.1s;
  -webkit-transition: transform 0.1s;
  transition: transform 0.1s;
}

.emotion-1:active:not([disabled]) {
  -webkit-transform: translateY(3px);
  -ms-transform: translateY(3px);
  transform: translateY(3px);
}

.emotion-0 {
  font-family: Roboto;
  color: #FFFFFF;
}

<button
    class="emotion-1"
    type="button"
  >
    <span
      class="emotion-0"
    >
      Submit
    </span>
  </button>
</DocumentFragment>
`;
